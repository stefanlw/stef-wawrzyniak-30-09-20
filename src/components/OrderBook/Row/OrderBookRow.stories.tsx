import React from 'react';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { OrderBookRow } from '.';
import { Text, View } from 'react-native';
import { spacing } from '~/theme/spacing';

storiesOf('Row', module)
  .add('single', () => (
    <View style={centre}>
      <OrderBookRow>
        <Text> {text('Column 1', 'Price')}</Text>
        <Text> {text('Column 2', 'Size')}</Text>
        <Text> {text('Column 3', 'Total')}</Text>
      </OrderBookRow>
    </View>
  ))
  .add('multiple', () => (
    <View style={centre}>
      <OrderBookRow>
        <Text> {text('Column 1', 'Price')}</Text>
        <Text> {text('Column 2', 'Size')}</Text>
        <Text> {text('Column 3', 'Total')}</Text>
      </OrderBookRow>
      <OrderBookRow>
        <Text> {text('Column 4', 'Column 4')}</Text>
        <Text> {text('Column 5', 'Column 5')}</Text>
        <Text> {text('Column 6', 'Column 6')}</Text>
      </OrderBookRow>
      <OrderBookRow>
        <Text> {text('Column 7', 'Column 7')}</Text>
        <Text> {text('Column 8', 'Column 8')}</Text>
        <Text> {text('Column 9', 'Column 9')}</Text>
      </OrderBookRow>
    </View>
  ));

const centre = {
  flex: 1,
  alignItems: 'center',
  marginHorizontal: spacing.xlarge,
};
