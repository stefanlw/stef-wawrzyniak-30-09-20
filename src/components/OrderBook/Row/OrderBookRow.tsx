import * as React from 'react';
import { View, StyleSheet, StyleProp, ViewStyle } from 'react-native';
import { Depth } from '~/features/OrderBook/types/OrderBookTypes';
import { Dimensions } from 'react-native';
import { colors } from '~/theme/colors';

export enum OrderType {
  ASK,
  BID,
}

type Props = {
  children: React.ReactNode[] | React.ReactNode;
  containerStyle?: StyleProp<ViewStyle>;
  depth?: Depth;
  orderType?: OrderType;
};

export const ORDER_BOOK_ROW_HEIGHT = 25;

const DEVICE_WIDTH = Dimensions.get('window').width;

export const OrderBookRow = ({
  children,
  containerStyle,
  depth,
  orderType,
}: Props) => (
  <View style={[styles.container, containerStyle]}>
    {!!depth && (
      <View
        style={[
          styles.depth,
          {
            backgroundColor:
              orderType === OrderType.ASK ? colors.red2 : colors.green2,
            width: DEVICE_WIDTH * depth,
          },
        ]}
      />
    )}
    {React.Children.map(children, child => (
      <Column style={React.Children.count(children) === 1 && styles.center}>
        {child}
      </Column>
    ))}
  </View>
);

type ColumnProps = {
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
};

const Column = ({ children, style }: ColumnProps) => (
  <View style={[styles.column, style]}>{children}</View>
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    height: ORDER_BOOK_ROW_HEIGHT,
    paddingHorizontal: 40,
  },
  column: {
    flex: 3,
    alignItems: 'flex-end',
  },
  center: {
    alignItems: 'center',
  },
  depth: { position: 'absolute', height: ORDER_BOOK_ROW_HEIGHT },
});
