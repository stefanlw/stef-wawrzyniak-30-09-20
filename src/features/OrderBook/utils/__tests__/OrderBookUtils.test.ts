import { OrderType } from '~/components/OrderBook/Row/OrderBookRow';
import {
  ASKS_BUFFER_MOCK,
  BIDS_BUFFER_MOCK,
  MOCK_ORDER_BOOK_STATE,
  ASKS_BUFFER_ENTRIES_MOCK,
} from '../../../../../__mocks__/OrderBook';

import {
  buildOrderBook,
  buildOrderBookLevels,
  calculateSpread,
  mergeOrders,
  getSortedBufferEntries,
} from '../OrderBookUtils';

describe('getSortedBufferEntries', () => {
  it('should return asks in the correct order', () => {
    const result = getSortedBufferEntries(ASKS_BUFFER_MOCK, OrderType.ASK);
    expect(result).toMatchSnapshot();
  });

  it('should return bids in the correct order', () => {
    const result = getSortedBufferEntries(BIDS_BUFFER_MOCK, OrderType.BID);
    expect(result).toMatchSnapshot();
  });
});

describe('buildOrderBookLevels', () => {
  it('should return array of Levels type with totals calculated', () => {
    const result = buildOrderBookLevels(ASKS_BUFFER_ENTRIES_MOCK);
    // TODO: use mocks instead of snapshots
    expect(result).toMatchSnapshot();
  });
});

describe('mergeOrders', () => {
  describe('given delete deltas', () => {
    it('should return the buffer without those prices', () => {
      const deleteDeltas: [number, number][] = [
        [10, 0],
        [15, 0],
      ];
      const buffer = {
        10: 20,
        15: 25,
        20: 30,
      };
      const result = mergeOrders(deleteDeltas, buffer);
      expect(result).toEqual({
        20: 30,
      });
    });
  });

  describe('given new level deltas', () => {
    it('should return the buffer with the new prices', () => {
      const newLevelDeltas: [number, number][] = [
        [35, 100],
        [45, 50],
      ];
      const buffer = { 25: 10, 50: 5 };
      const result = mergeOrders(newLevelDeltas, buffer);
      expect(result).toEqual({ 25: 10, 35: 100, 45: 50, 50: 5 });
    });
  });
});

describe('calculateSpread', () => {
  it('should return correct Spread object', () => {
    expect(
      calculateSpread(MOCK_ORDER_BOOK_STATE.asks, MOCK_ORDER_BOOK_STATE.bids),
    ).toEqual({ percentage: '0.00', value: '6.0' });
  });
});

// TODO: fix
// const mockGetSortedBufferEntries = jest.fn()
// const mockBuildOrderBookLevels = jest.fn()
// jest.mock('../OrderBookUtils', () => ({
//   ...jest.requireActual('../OrderBookUtils'),
//   getSortedBufferEntries: mockGetSortedBufferEntries,
//   buildOrderBookLevels: mockBuildOrderBookLevels
// }))

// describe('buildOrderBook', () => {
//     it ('should call getSortedBufferEntries', () => {
//     buildOrderBook(MERGE_ORDERS_ASKS_RESULT_MOCK, MERGE_ORDERS_BIDS_RESULT_MOCK)
//     expect(mockGetSortedBufferEntries).toHaveBeenCalled()
//   })
// });
