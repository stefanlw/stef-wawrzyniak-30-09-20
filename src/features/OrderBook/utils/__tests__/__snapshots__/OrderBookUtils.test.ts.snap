// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOrderBookLevels should return array of Levels type with totals calculated 1`] = `
Array [
  Object {
    "price": "48384",
    "size": 1,
    "total": 1,
  },
  Object {
    "price": "48388",
    "size": 4725,
    "total": 4726,
  },
  Object {
    "price": "48390",
    "size": 1235,
    "total": 5961,
  },
  Object {
    "price": "48391",
    "size": 15000,
    "total": 20961,
  },
  Object {
    "price": "48392",
    "size": 57158,
    "total": 78119,
  },
  Object {
    "price": "48398",
    "size": 450,
    "total": 78569,
  },
  Object {
    "price": "48401",
    "size": 2607,
    "total": 81176,
  },
  Object {
    "price": "48407",
    "size": 9681,
    "total": 90857,
  },
  Object {
    "price": "48412",
    "size": 2092,
    "total": 92949,
  },
  Object {
    "price": "48414",
    "size": 25000,
    "total": 117949,
  },
  Object {
    "price": "48416",
    "size": 53820,
    "total": 171769,
  },
  Object {
    "price": "48419",
    "size": 7500,
    "total": 179269,
  },
  Object {
    "price": "48376.5",
    "size": 20000,
    "total": 199269,
  },
  Object {
    "price": "48381.5",
    "size": 24072,
    "total": 223341,
  },
  Object {
    "price": "48388.5",
    "size": 1,
    "total": 223342,
  },
  Object {
    "price": "48391.5",
    "size": 6922,
    "total": 230264,
  },
  Object {
    "price": "48392.5",
    "size": 86628,
    "total": 316892,
  },
  Object {
    "price": "48395.5",
    "size": 1000,
    "total": 317892,
  },
  Object {
    "price": "48397.5",
    "size": 4479,
    "total": 322371,
  },
  Object {
    "price": "48398.5",
    "size": 8338,
    "total": 330709,
  },
  Object {
    "price": "48400.5",
    "size": 5000,
    "total": 335709,
  },
  Object {
    "price": "48408.5",
    "size": 2500,
    "total": 338209,
  },
  Object {
    "price": "48418.5",
    "size": 4578,
    "total": 342787,
  },
  Object {
    "price": "48421.5",
    "size": 12222,
    "total": 355009,
  },
  Object {
    "price": "48423.5",
    "size": 7522,
    "total": 362531,
  },
]
`;

exports[`getSortedBufferEntries should return asks in the correct order 1`] = `
Array [
  Array [
    "48376.5",
    20000,
  ],
  Array [
    "48381.5",
    24072,
  ],
  Array [
    "48384",
    1,
  ],
  Array [
    "48388",
    4725,
  ],
  Array [
    "48388.5",
    1,
  ],
  Array [
    "48390",
    1235,
  ],
  Array [
    "48391",
    15000,
  ],
  Array [
    "48391.5",
    6922,
  ],
  Array [
    "48392",
    57158,
  ],
  Array [
    "48392.5",
    86628,
  ],
  Array [
    "48395.5",
    1000,
  ],
  Array [
    "48397.5",
    4479,
  ],
  Array [
    "48398",
    450,
  ],
  Array [
    "48398.5",
    8338,
  ],
  Array [
    "48400.5",
    5000,
  ],
  Array [
    "48401",
    2607,
  ],
  Array [
    "48407",
    9681,
  ],
  Array [
    "48408.5",
    2500,
  ],
  Array [
    "48412",
    2092,
  ],
  Array [
    "48414",
    25000,
  ],
  Array [
    "48416",
    53820,
  ],
  Array [
    "48418.5",
    4578,
  ],
  Array [
    "48419",
    7500,
  ],
  Array [
    "48421.5",
    12222,
  ],
  Array [
    "48423.5",
    7522,
  ],
]
`;

exports[`getSortedBufferEntries should return bids in the correct order 1`] = `
Array [
  Array [
    "48361.5",
    1000,
  ],
  Array [
    "48358.5",
    2500,
  ],
  Array [
    "48358",
    4827,
  ],
  Array [
    "48354.5",
    6848,
  ],
  Array [
    "48350",
    9800,
  ],
  Array [
    "48348",
    1675,
  ],
  Array [
    "48346",
    4651,
  ],
  Array [
    "48345.5",
    5000,
  ],
  Array [
    "48343",
    25000,
  ],
  Array [
    "48341",
    206,
  ],
  Array [
    "48340",
    22902,
  ],
  Array [
    "48338.5",
    63160,
  ],
  Array [
    "48338",
    20000,
  ],
  Array [
    "48337.5",
    5748,
  ],
  Array [
    "48337",
    9267,
  ],
  Array [
    "48336.5",
    2500,
  ],
  Array [
    "48334",
    79726,
  ],
  Array [
    "48333",
    20000,
  ],
  Array [
    "48330",
    7758,
  ],
  Array [
    "48328",
    5000,
  ],
  Array [
    "48327",
    4729,
  ],
  Array [
    "48326.5",
    2500,
  ],
  Array [
    "48322",
    9660,
  ],
  Array [
    "48321.5",
    3053,
  ],
  Array [
    "48320.5",
    240000,
  ],
]
`;
